sys/kern/kern_malloc.c

```diff
@@ -508,7 +511,10 @@ contigmalloc_domainset(unsigned long size, struct malloc_type *type,
 void
 contigfree(void *addr, unsigned long size, struct malloc_type *type)
 {
-	CHERI_ASSERT_VALID(addr);
+
+#ifdef __CHERI_PURE_CAPABILITY__
+	KASSERT(cheri_gettag(addr), ("Expected valid capability"));
```

Comment:
I'm not convinced that adding `#ifdef`s instead of using macros makes this more readable.

If the CHERI_* macros are considered too ugly (or easy to misuse because it isn't obvious that it's a no-op for hybrid?), how about  `KASSERT_PURECAP()` to avoid all the `#ifdef __CHERI_PURE_CAPABILITY__`

Reply:
I disagree, this makes it extremely clear, more than any magic wrapper macro could ever do.

Reply:
True, it makes it more clear but also more ugly and makes it easier to make mistakes such as `#ifdef __CHERI_PURE_CAPABILTY__`, but it seems like only I make those mistakes :(
```
libexec/rtld-elf/mips/reloc.c:#endif  /* !defined(__CHERI_PURE_CAPABILITY) */
```
Fortunately it's just in a comment.

Anyway I don't really mind either way, but any KASSERT-related macros should be in sys/system.h rather than cheri/cheric.h

Reply:
I think using the KASSERT directly makes the fact that those will assert only in the purecap kernel clearer. Also the assertions are not very common and generally used only to double check on allocators behaviour. There are no KASSERT-related macros in cheric.h anymore at this point.